sdk/python/kfp/compiler/compiler.py

```diff
@@ -678,6 +678,11 @@ def _create_dag_templates(self, pipeline, op_transformers=None, op_to_templates_
                             pipeline_name=self._pipeline_name_param,
                             pipeline_root=self._pipeline_root_param,
                             launcher_image=self._launcher_image)
+
+      if "Executor executes v2-based Python function components." in op.command[3] and\
```

Comment:
I tried to find a better solution for this, but there dont seems to be any variable indication when it is v2 '@component', WDYT @chensun?

Reply:
Right, this seems to be quite hacky and unreliable. 

> but there dont seems to be any variable indication when it is v2 '@component'

How about we add such an indicator variable when `@component` is used? WDYT?

Reply:
Yes, that makes sense to me. A possible solution would be to set:
https://github.com/kubeflow/pipelines/blob/b57bbeeb982744556ab79a132de5d5fd0c7c0c82/sdk/python/kfp/components/_structures.py#L367

here:https://github.com/kubeflow/pipelines/blob/b57bbeeb982744556ab79a132de5d5fd0c7c0c82/sdk/python/kfp/components/_python_op.py#L540

it is a bit unclear for me what the version argument are used to describe today and is adding a `google.com/cloud/pipelines/component/v2` is the right usage here. What are your thoughts @chensun?

Reply:
The current plan is to have v2 be a hermetic namespace at some point (even if that means duplication of code across v1 and v2). So I don't think this is the right solution. How about, if v2 components are used, the v1 compiler should auto-compile in v2-compatible mode?

Reply:
> The current plan is to have v2 be a hermetic namespace at some point...

Could you elaborate on this @neuromage? What does this mean?

> How about, if v2 components are used, the v1 compiler should auto-compile in v2-compatible mode?

I would prefer to keep the compiler selected by the users, I think it will be unclear for users(especially new users that are more prone to make this misstake) if the compiler are switched based upon the components.

Reply:
> > The current plan is to have v2 be a hermetic namespace at some point...
> 
> Could you elaborate on this @neuromage? What does this mean?

It means the v2 namespace (`kfp.v2.*`) will be self-contained, so we can avoid mix-and-match of concepts from v1 and v2.

>
> > How about, if v2 components are used, the v1 compiler should auto-compile in v2-compatible mode?
> 
> I would prefer to keep the compiler selected by the users, I think it will be unclear for users(especially new users that are more prone to make this misstake) if the compiler are switched based upon the components.

I disagree. The compiler mode is an implementation detail. The user doesn't need to know v2 requires a different compilation mode.

Reply:
@chensun What are your thoughts on this?

Reply:
@neuromage The following are known breaking changes of v2 compatible mode (some are planned to be resolved, but not all):
* https://github.com/kubeflow/pipelines/issues/5834 
* https://github.com/kubeflow/pipelines/issues/5831
* https://github.com/kubeflow/pipelines/issues/5711
* https://github.com/kubeflow/pipelines/issues/5673
* https://github.com/kubeflow/pipelines/issues/6132
* v2 compatible mode will require KFP backend 1.7.0+, but v1 mode does not have any prerequisites to KFP backend
* v2 compatible mode and v1 mode have different perf / scalability characteristics, which we haven't verified in large scale.
* Because there aren't enough usage, it's very likely there are more changes we haven't found/thought of.

> The user doesn't need to know v2 requires a different compilation mode.

I think you'll only be able to say this, until most of the above issues are resolved. Do you think that's the direction we will target?

Reply:
I think tracking all the breaking changes in one place is very useful, just filed https://github.com/kubeflow/pipelines/issues/6133

Reply:
> @chensun What are your thoughts on this?

I think there might be some challenges to alter compilation mode based on feature usage, given this decorator isn't the only feature not compatible with v1. If we were to have the pattern of changing compilation mode based on usage. There would be more than one cases, and it would make it hard for users to understand what contributed to the compilation mode change. 
So I think throwing an error would be okay for now, and we can revisit this later? @neuromage WDYT?

Reply:
ping @neuromage

Reply:
I'm ok to throw an error when compiling in v1 mode, but can we should ideally have a better way to detect this rather than looking at the docstrings. I think @chensun is planning for `@component` decorator to return a v2 class called `BaseComponent` so we can check for this and throw the error more reliably.